{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Finding novel insertions\n",
    "\n",
    "\n",
    "Environment: bioconda's `sambamba samtools pigz parallel bxtools blast snakemake spades bioawk` + `supernova` and `ema` (installed in extras)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run longranger to get `NA12878_WGS_v2_phased_possorted_bam.bam` or `COLO829-10x_phased_possorted_bam.bam`\n",
    "\n",
    "```\n",
    "cd /data/cephfs/punim0010/projects/Saveliev_10X/COLO829-10x/insertions\n",
    "```\n",
    "\n",
    "Extract unmapped reads or reads with unmapped mate:\n",
    "\n",
    "```\n",
    "mkdir unmapped_or_mate_is_unmapped\n",
    "\n",
    "NA12878:\n",
    "sambamba view -f bam -F \"unmapped or mate_is_unmapped\" -t 20 /data/cephfs/punim0010/projects/Saveliev_10X/NA12878-10x/NA12878_WGS_v2_phased_possorted_bam.bam | sambamba sort -n -t 20 -m 100G /dev/stdin --show-progress > unmapped_or_mate_is_unmapped/NA12878_WGS.bam\n",
    "\n",
    "COLO829:\n",
    "sambamba view -f bam -F \"unmapped or mate_is_unmapped\" -t 20 COLO829-10x_phased_possorted_bam.bam | sambamba sort -n -t {threads} -m 100G /dev/stdin --show-progress > unmapped_or_mate_is_unmapped/COLO829-10x.bam\n",
    "```\n",
    "\n",
    "Split by BX tag:\n",
    "\n",
    "```\n",
    "cd unmapped_or_mate_is_unmapped\n",
    "mkdir bx_split\n",
    "bxtools split NA12878_WGS.bam -a bx_split/NA12878_WGS -m 1000 > bx_split/NA12878_WGS_bx_counts.txt\n",
    "bxtools split COLO829-10x.bam -a bx_split/COLO829-10x -m 1000 > bx_split/COLO829-10x_bx_counts.txt\n",
    "```\n",
    "\n",
    "Extract tags with at least 1000 reads:\n",
    "\n",
    "```\n",
    "sort -k 2 -r -n NA12878_WGS_bx_counts.txt | py -x \"x if int(x.split()[1]) >= 1000 else None\" > NA12878_WGS_bx_counts_cnt1000.txt\n",
    "sort -k 2 -r -n COLO829-10x_bx_counts.txt | py -x \"x if int(x.split()[1]) >= 1000 else None\" > COLO829-10x_bx_counts_cnt1000.txt\n",
    "```\n",
    "\n",
    "Count all reads in full BAM by BX:\n",
    "\n",
    "```\n",
    "bxtools stats /data/cephfs/punim0010/projects/Saveliev_10X/NA12878-10x/NA12878_WGS_v2_phased_possorted_bam.bam > /data/cephfs/punim0010/projects/Saveliev_10X/NA12878-10x/qc/NA12878_WGS_v2_phased_possorted_bam.bxstats.txt\n",
    "\n",
    "bxtools stats /data/cephfs/punim0010/projects/Saveliev_10X/NA12878-10x/NA12878_WGS_v2_phased_possorted_bam.bam > /data/cephfs/punim0010/projects/Saveliev_10X/NA12878-10x/qc/NA12878_WGS_v2_phased_possorted_bam.bxstats.txt\n",
    "\n",
    "```\n",
    "\n",
    "Extract BX with many unmapped:\n",
    "\n",
    "```\n",
    "grep -f <(cut -f1 NA12878_WGS_bx_counts_cnt1000.txt) /data/cephfs/punim0010/projects/Saveliev_10X/NA12878-10x/qc/NA12878_WGS_v2_phased_possorted_bam.bxstats.txt > NA12878_WGS_full_bx_stats_cnt1000.txt\n",
    "```\n",
    "\n",
    "------------------------\n",
    "Optionally, we can split full BAM by BX tag:\n",
    "\n",
    "```\n",
    "bxtools split NA12878_WGS_v2_phased_possorted_bam.bam -a bx_split/NA12878_WGS -m 1000 > bx_split/NA12878_WGS_bx_counts.txt\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as pt\n",
    "from pprint import pprint\n",
    "from collections import namedtuple\n",
    "\n",
    "bxs = dict()\n",
    "\n",
    "with open('files/NA12878_WGS_bx_counts_unmapped_or_mate_top.txt') as f:\n",
    "    for l in f:\n",
    "        l = l.strip()\n",
    "        if l:\n",
    "            bx, cnt = l.split()\n",
    "            cnt = int(cnt)\n",
    "            if cnt >= 1000:\n",
    "                bxs[bx] = {'Or mate': int(cnt)}\n",
    "\n",
    "with open('files/NA12878_WGS_bx_counts_unmapped_top.txt') as f:\n",
    "    for l in f:\n",
    "        l = l.strip()\n",
    "        if l:\n",
    "            bx, cnt = l.split()\n",
    "            if bx in bxs:\n",
    "                bxs[bx]['Unaln'] = int(cnt)\n",
    "    \n",
    "# counts = [c for bx, c in bx_cnt.items() if c > 500]\n",
    "# print(len(counts))\n",
    "# pt.plot(counts)\n",
    "\n",
    "with open('files/NA12878_WGS_full_bx_stats_cnt1000.txt') as f:\n",
    "    for l in f:\n",
    "        l = l.strip()\n",
    "        if l:\n",
    "            bx, cnt, mis, mmq, mas = l.split()\n",
    "            if bx in bxs:\n",
    "                bxs[bx]['Aln'] = int(cnt)\n",
    "                bxs[bx]['IS'] = int(mis)\n",
    "                bxs[bx]['MQ'] = int(mmq)\n",
    "                bxs[bx]['AS'] = int(mas)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22\n",
      "                     Aln  Unaln | Mate unaln  Unaln  Aln & Mate unaln   IS  MQ\n",
      "CATGTAGAGTCGTACT-1  4037              94.70% 94.43%                11  237  60\n",
      "GCCTGTTAGCCCGTGT-1  1342              86.44% 85.77%                 9  348  60\n",
      "ACTGTAGGTTCGTAAC-1  1647              84.03% 83.79%                 4  347  60\n",
      "CATGTAGGTAGATGTA-1  1250              84.00% 83.28%                 9   19  19\n",
      "GCCTGTTGTTCGCGGT-1  1921              73.30% 72.98%                 6  348  60\n",
      "CATGTAGAGAGTCTCT-1  1828              72.87% 72.10%                14   20  15\n",
      "GCCTGTTGTCTTACCC-1  1827              72.03% 71.65%                 7  374  60\n",
      "GCCTGTTAGTCGAGTG-1  1515              69.37% 68.84%                 8  349  60\n",
      "TCAGTAGGTCATGTAC-1  2070              69.47% 68.41%                22  304  60\n",
      "GCCTGTTAGCTCCGTG-1  1714              68.14% 67.74%                 7  368  60\n",
      "ACTGTAGAGGTAAACT-1  1507              67.09% 66.42%                10   20  19\n",
      "GCCTGTTAGGCCCTCA-1  3324              63.96% 63.45%                17  308  60\n",
      "CATGTAGGTTGATCAC-1  1649              63.61% 62.77%                14  315  60\n",
      "GCCTGTTGTTTCGTTT-1  1933              63.68% 62.44%                24  322  60\n",
      "GCCTGTTAGTCGGCTC-1  2126              61.90% 61.43%                10  339  60\n",
      "GCCTGTTGTTTGGCTA-1  1921              64.29% 56.95%               141  314  60\n",
      "TGCGCAGGTACTGTAC-1  2327              56.81% 55.82%                23  334  60\n",
      "GCCTGTTGTCCGGGTT-1  2053              52.70% 50.02%                55  288  49\n",
      "GCCTGTTGTACGTTTG-1  2316              50.69% 50.00%                16  286  60\n",
      "AAGCCCGCATCCCACT-1  2524              41.32% 37.08%               107  264  20\n",
      "GTGCCTTCAGATCTGT-1  7461              21.61% 20.68%                69  336  60\n",
      "GTTGCAACAGATCTGT-1  5943              19.60% 15.01%               273  329  60\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df = pd.DataFrame.from_dict(bxs, orient='index')\n",
    "print(len(df))\n",
    "df['Aln & Mate unaln'] = df['Unaln | Mate unaln'] - df['Unaln']\n",
    "\n",
    "df['Unaln | Mate unaln'] = 100.0 * df['Unaln | Mate unaln'] / df['Aln']\n",
    "df['Unaln'] = 100.0 * df['Unaln'] / df['Aln']\n",
    "#df['Aln & Mate unaln'] = 100.0 * df['Aln & Mate unaln'] / df['Aln']\n",
    "\n",
    "df = df.sort_values(by='Unaln', ascending=False)\n",
    "\n",
    "pd.options.display.float_format = '{:.2f}%'.format\n",
    "#print('\\n'.join(df.index))\n",
    "print(df[['Aln', 'Unaln | Mate unaln', 'Unaln', 'Aln & Mate unaln', 'IS', 'MQ']])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each tag, extract mapped reads with unmapped mates:\n",
    "\n",
    "```python\n",
    "!cd /data/cephfs/punim0010/projects/Saveliev_10X/COLO829-10x/insertions/unmapped_or_mate_is_unmapped/split_by_bx/\n",
    "\n",
    "for tag in df.index:\n",
    "    !sambamba view -F \"not unmapped and mapping_quality >= 60\" NA12878_WGS.{tag}-1.bam -f bam | samtools sort > NA12878_WGS.{tag}-1.mapped.bam\n",
    "```\n",
    "\n",
    "Manually reviewing the reads, we can find one BX with mapped reads spanning consistently more or less one region (possibly breakpoint?)\n",
    "\n",
    "```\n",
    "samtools view NA12878_WGS.GCCTGTTGTTTCGTTT-1.mapped.bam\n",
    "\n",
    "ST-E00273:259:H7WY3ALXX:1:1217:21207:15479  121 chr2  71844818  60  73M1D11M2I30M4D12M  chr7  128046340  0\n",
    "ST-E00273:259:H7WY3ALXX:1:1220:8156:43677    73 chr2  71844971  60  47M6I7M2I66M        chr10    654413  0\n",
    "ST-E00273:259:H7WY3ALXX:2:1118:11373:44152   73 chr2  71845284  60  52M13I2M1D61M       chr21  44035538  0\n",
    "ST-E00273:259:H7WY3ALXX:1:2121:15006:67814   89 chr2  71845342  60  14S54M7I53M         =      33141434  0\n",
    "ST-E00273:259:H7WY3ALXX:3:2117:1499:37664    89 chr2  71845402  60  77M2I22M1I17M1D9M   =      71845700  0\n",
    "ST-E00273:259:H7WY3ALXX:1:1112:18913:57038  169 chr2  71845409  60  7M2I66M76S          *             0  0\n",
    "ST-E00273:259:H7WY3ALXX:1:2124:12134:1977    73 chr2  71845440  60  79M2D21M1D24M4S     chr20  25048324  0\n",
    "ST-E00273:259:H7WY3ALXX:2:1106:10602:14828  153 chr2  71845521  60  30S25M2I40M1D13M41S *             0  0\n",
    "ST-E00273:259:H7WY3ALXX:2:1105:1458:67937    89 chr2  71845525  60  35S93M              chr4   72052386  0\n",
    "ST-E00273:259:H7WY3ALXX:1:2203:29985:65001   89 chr2  71845588  60  9S78M2D29M2D12M     =      71844822  0\n",
    "ST-E00273:259:H7WY3ALXX:3:2207:11921:48230  153 chr2  71845590  60  75S54M7I15M         *             0  0\n",
    "ST-E00273:259:H7WY3ALXX:1:2120:18832:35801   89 chr2  71845617  60  46M2D79M3S          =      71845704  0\n",
    "ST-E00273:259:H7WY3ALXX:1:2204:30888:68430   73 chr2  71845848  60  6S86M36S            hs37d5 31586471  0\n",
    "ST-E00273:259:H7WY3ALXX:1:2204:18913:24198   73 chr4  53065963  60  23M4I19M3D38M44S    chr15  79384471  0\n",
    "ST-E00273:259:H7WY3ALXX:1:1104:28270:7181   121 chr14 32925056  60  128M                chr1   30986160  0\n",
    "ST-E00273:259:H7WY3ALXX:2:1101:22110:57899   73 chr14 32966430  60  128M                chr13  52303476  0\n",
    "ST-E00273:259:H7WY3ALXX:2:2103:30452:36311   89 chr14 33028682  60  128M                chr18  71871932  0\n",
    "ST-E00273:259:H7WY3ALXX:3:2201:23977:47316  105 chr21 28080947  60  128M                chr5   77330570  0\n",
    "```\n",
    "\n",
    "Now get all reads for that tag to see if they localized at the same region, then we'll see it in IGV:\n",
    "\n",
    "```\n",
    "cd /data/cephfs/punim0010/projects/Saveliev_10X/NA12878-10x\n",
    "bamtools filter -tag BX:GCCTGTTGTTTCGTTT-1 -in NA12878_WGS_v2_phased_possorted_bam.bam -out NA12878_WGS_v2_phased_possorted_bam-GCCTGTTGTTTCGTTT-1.bam\n",
    "```\n",
    "\n",
    "Extract that region to download locally on IGV:\n",
    "\n",
    "```\n",
    "sambamba slice NA12878_WGS_v2_phased_possorted_bam.bam chr2:71844818-71847848 -o NA12878_WGS_v2_phased_possorted_chr2_breakpoint.bam\n",
    "```\n",
    "\n",
    "Extracting this one to for comparison:\n",
    "\n",
    "```\n",
    "sambamba slice NA12878_WGS_v2_phased_possorted_bam.bam chr19:33788840-33791956 -o NA12878_WGS_v2_phased_possorted_CEBPA.bam\n",
    "\n",
    "```\n",
    "\n",
    "In the large SVs file downloaded from the 10x website, we can find those SVs called with LOWQ: `/data/cephfs/punim0010/projects/Saveliev_10X/NA12878-10x/longranger/downloaded/NA12878_WGS_v2_large_svs.vcf.gz`\n",
    "\n",
    "```\n",
    "chr2   72650821  call_2031    A  <INV>                 5  LOWQ  END=72704553;CIEND=0,1;CIPOS=0,1;SVTYPE=INV;SVLEN=53732;PS=71845775;HAP_ALLELIC_FRAC=0.25;ALLELIC_FRAC=0.0200501253133;PAIRS=1;SPLIT=0;WildCov=.;MolTtl=.;MolTtlNoR=.;MolDel=.\n",
    "chr2   72790392  call_3830    A  <DUP:TANDEM>          3  LOWQ  END=72858227;CIEND=0,1;CIPOS=0,1;SVTYPE=DUP;SVLEN=67835;PS=71845775;HAP_ALLELIC_FRAC=0.230769230769;ALLELIC_FRAC=0.00617283950617;PAIRS=1;SPLIT=0;WildCov=.;MolTtl=.;MolTtlNoR\n",
    "chr2   73124465  call_2875    G  <INV>                 4  LOWQ  END=73195439;CIEND=-24,25;CIPOS=-24,25;SVTYPE=INV;SVLEN=70974;PS=71845775;HAP_ALLELIC_FRAC=0.153846153846;ALLELIC_FRAC=0.0128755364807;PAIRS=2;SPLIT=0;WildCov=.;MolTtl=.;MolT\n",
    "chr2   73877193  call_2032    A  <UNK>                 5  LOWQ  END=73940170;CIEND=-8,8;CIPOS=-6,6;SVTYPE=UNK;SVLEN=62977;PS=71845775;HAP_ALLELIC_FRAC=0.172413793103;ALLELIC_FRAC=0.0178173719376;PAIRS=3;SPLIT=0;WildCov=.;MolTtl=.;MolTtlNo\n",
    "chr2   76051830  call_840     C  <INV>                 8  LOWQ  END=76110415;CIEND=0,1;CIPOS=0,1;SVTYPE=INV;SVLEN=58585;PS=71845775;HAP_ALLELIC_FRAC=0.228571428571;ALLELIC_FRAC=0.0220264317181;PAIRS=1;SPLIT=0;WildCov=.;MolTtl=.;MolTtlNoR=\n",
    "chr2   76213102  call_2033    A  <INV>                 5  LOWQ  END=76259489;CIEND=0,1;CIPOS=0,1;SVTYPE=INV;SVLEN=46387;PS=71845775;HAP_ALLELIC_FRAC=0.192307692308;ALLELIC_FRAC=0.0216216216216;PAIRS=1;SPLIT=0;WildCov=.;MolTtl=.;MolTtlNoR=\n",
    "chr2   78278376  call_3831    T  <INV>                 3  LOWQ  END=78322534;CIEND=0,1;CIPOS=0,1;SVTYPE=INV;SVLEN=44158;PS=71845775;HAP_ALLELIC_FRAC=0.103448275862;ALLELIC_FRAC=0.0113314447592;PAIRS=1;SPLIT=0;WildCov=.;MolTtl=.;MolTtlNoR=\n",
    "chr2   78415466  call_2876    A  <UNK>                 4  LOWQ  END=78476352;CIEND=-9,9;CIPOS=-9,9;SVTYPE=UNK;SVLEN=60886;PS=71845775;HAP_ALLELIC_FRAC=0.137931034483;ALLELIC_FRAC=0.0176767676768;PAIRS=3;SPLIT=0;WildCov=.;MolTtl=.;MolTtlNo\n",
    "chr2   78887169  call_3832    G  <INV>                 3  LOWQ  END=78950399;CIEND=0,1;CIPOS=0,1;SVTYPE=INV;SVLEN=63230;PS=71845775;HAP_ALLELIC_FRAC=0.111111111111;ALLELIC_FRAC=0.00973236009732;PAIRS=1;SPLIT=0;WildCov=.;MolTtl=.;MolTtlNoR\n",
    "chr2   79106744  call_3833    A  <UNK>                 3  LOWQ  END=79159256;CIEND=-30,31;CIPOS=-30,31;SVTYPE=UNK;SVLEN=52512;PS=71845775;HAP_ALLELIC_FRAC=0.2;ALLELIC_FRAC=0.0171568627451;PAIRS=2;SPLIT=0;WildCov=.;MolTtl=.;MolTtlNoR=.;Mol\n",
    "chr2   80313256  call_1487    T  <INV>                 6  LOWQ  END=80353748;CIEND=-15,15;CIPOS=-15,15;SVTYPE=INV;SVLEN=40492;PS=71845775;HAP_ALLELIC_FRAC=0.25;ALLELIC_FRAC=0.0263157894737;PAIRS=2;SPLIT=0;WildCov=.;MolTtl=.;MolTtlNoR=.;Mo\n",
    "```\n",
    "\n",
    "Viewing in IGV: `chr2:71,844,732-71,845,762`\n",
    "\n",
    "![alt text](img/NA12878_WGS_v2_phased_possorted_chr2_IGV.png)\n",
    "\n",
    "\"Show BX\" mode:\n",
    "![alt text](img/NA12878_WGS_v2_phased_possorted_chr2_IGV_showBX.png)\n",
    "\n",
    "For comparison, this is how a regular region looks:\n",
    "\n",
    "![alt text](img/NA12878_CEBPA_igv_snapshot.png)\n",
    "\n",
    "![alt text](img/NA12878_CEBPA_igv_snapshot_viewBX.png)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Try de novo assembly of unmapped reads with Supernova\n",
    "\n",
    "Extract reads:\n",
    "\n",
    "```\n",
    "bedtools bamtofastq -i COLO829-10x.bam -fq COLO829-10x_S1_L001_R1_001.fastq -fq2 COLO829-10x_S1_L001_R2_001.fastq 2>COLO829-10x.unpaired\n",
    "bedtools bamtofastq -i NA12878_WGS.bam -fq NA12878_WGS_S1_L001_R1_001.fastq -fq2 NA12878_WGS_S1_L001_R2_001.fastq 2>NA12878_WGS.unpaired\n",
    "```\n",
    "\n",
    "Try assemble with supernova:\n",
    "\n",
    "```\n",
    "cd /g/data3/gx8/projects/Saveliev_10X/NA12878-10x/insertions/unmapped_or_mate_is_unmapped/assemble\n",
    "qsub run.sh\n",
    "```\n",
    "\n",
    "However, it fails with error:\n",
    "\n",
    "```\n",
    "[error] The fraction of input reads having valid barcodes is 1.68 pct, whereas the ideal is at least 80 pct.  This condition could have multiple causes including wrong library type, failed library construction and low sequence quality on the barcode bases.  This could have a severe effect on assembly performance, and Supernova has not been tested on data with these properties, so execution will be terminated.\n",
    "\n",
    "We observe only 36.67 pct of bases on read two with quality scores at least 30. Ideally, we expect at least 75 pct. Data quality issues of this type are difficult to diagnose, but might be caused by defects in sequencing reagents or sequencing instrument condition. Use of low quality reads usually reduces assembly quality.\n",
    "```\n",
    "\n",
    "The reason is that extracted reads do no preserve barcodes. We need to extract those reads from the original `fastq` files. To save computational time, we first want to look for particular retroviruses as play with only reads mapping to one virus.\n",
    "\n",
    "### Extract viral sequences\n",
    "\n",
    "Searching viral sequences using 3 different approaches.\n",
    "\n",
    "- Use Mash (http://mash.readthedocs.io/en/latest/tutorials.html)\n",
    "\n",
    "```\n",
    "mash sketch /g/data3/gx8/extras/vlad/bcbio/genomes/Hsapiens/GRCh37/viral/gdc-viral.fa -i -o mash/gdc-viral\n",
    "mash screen gdc-viral.msh ../NA12878_WGS.merged.fq > screen_gdc.tab\n",
    "sort -gr screen_gdc.tab | head\n",
    "0.833811        22/1000 2       2.09944e-33     HPV20\n",
    "0.828011        19/1000 4       7.40488e-28     HPV6\n",
    "0.821262        16/1000 3       1.62953e-22     HPV83\n",
    "0.816057        14/1000 6       4.38911e-19     HPV114\n",
    "0.816057        14/1000 4       4.38911e-19     HPV19\n",
    "0.816057        14/1000 3       4.38911e-19     HPV77\n",
    "0.816057        14/1000 1       4.38911e-19     HPV87\n",
    "0.813182        13/1000 1       2.05593e-17     HPV29\n",
    "0.810088        12/1000 49      8.93544e-16     HPV12\n",
    "0.810088        12/1000 2       8.93544e-16     HPV71\n",
    "```\n",
    "\n",
    "Also against RefSeq:\n",
    "\n",
    "```\n",
    "mash screen refseq.genomes.k21s1000.msh ../NA12878_WGS.merged.fq > screen_refseq.tab\n",
    "```\n",
    "\n",
    "- Kraken against minikraken (http://ccb.jhu.edu/software/kraken/MANUAL.html)\n",
    "\n",
    "```\n",
    "kraken --preload --db /g/data3/gx8/extras/kraken/minikraken_20171019_8GB --fastq-input fastq/NA12878_WGS.R1.fq --threads 9 --out kraken/kraken_out --min-hits 2 --quick\n",
    "```\n",
    "\n",
    "- Align against the viral database:\n",
    "\n",
    "```\n",
    "bwa mem -t 9 /g/data3/gx8/extras/vlad/bcbio/genomes/Hsapiens/GRCh37/viral/gdc-viral.fa NA12878_WGS.R1.fq NA12878_WGS.R2.fq | bamsort inputthreads=9 outputthreads=9 inputformat=sam index=1 indexfilename=viral_mapping/NA12878_viral.bam.bai O=viral_mapping/NA12878_viral.bam\n",
    "# Real time: 180.500 sec; CPU: 1553.289 sec\n",
    "\n",
    "samtools idxstats viral_mapping/NA12878_viral.bam | py -x \"'\\t'.join([x.split()[0], x.split()[1], x.split()[2], str(int(x.split()[2]) / int(x.split()[1]))]) if int(x.split()[1]) > 0 else None\" | sort -nr -k4,4 | head | cols\n",
    "\n",
    "name    mapped coverage\n",
    "HPV71   87280  10.8597735473\n",
    "HPV47   45913  5.94266114419\n",
    "HPV82   20185  2.56480304956\n",
    "HPV42   10600  1.33889099406\n",
    "HCV-1   11845  1.22797014306\n",
    "HCV-2   10985  1.1311914324\n",
    "HPV51    8678  1.11142418033\n",
    "HPV21    8375  1.07661653169\n",
    "HPV105   5904  0.770053475936\n",
    "HPV20    5500  0.70903699884\n",
    "```\n",
    "\n",
    "### Mapping to HPV71\n",
    "\n",
    "According to the viral mapping, the top hit is HPV71 with 10x coverage. Aligning reads to HPV71 only:\n",
    "\n",
    "```\n",
    "samtools faidx /g/data3/gx8/extras/vlad/bcbio/genomes/Hsapiens/GRCh37/viral/gdc-viral.fa HPV71 > viral_mapping/HPV71.fa\n",
    "\n",
    "bwa index viral_mapping/HPV71.fa\n",
    "\n",
    "bwa mem -t 9 viral_mapping/HPV71.fa NA12878_WGS.R1.fq NA12878_WGS.R2.fq | bamsort inputthreads=9 outputthreads=9 inputformat=sam index=1 indexfilename=viral_mapping/NA12878_HPV71.bam.bai O=viral_mapping/NA12878_HPV71.bam\n",
    "\n",
    "samtools idxstats viral_mapping/NA12878_HPV71.bam | py -x \"'\\t'.join([x.split()[0], x.split()[2], str(int(x.split()[2]) / int(x.split()[1])), str(int(x.split()[3]) / int(x.split()[1]))]) if int(x.split()[1]) > 0 else None\" | sort -nr -k3,3 | head | cols\n",
    "# HPV71  102573  12.7625979843  4.10240139355\n",
    "\n",
    "samtools stats viral_mapping/NA12878_HPV71.bam > viral_mapping/NA12878_HPV71.stats\n",
    "SN      bases mapped (cigar):   3815054 # more accurate\n",
    "SN      bases trimmed:  0\n",
    "SN      bases duplicated:       0\n",
    "SN      mismatches:     180824  # from NM fields\n",
    "SN      error rate:     4.739749e-02    # mismatches / bases mapped (cigar)\n",
    "SN      average length: 135\n",
    "SN      maximum length: 151\n",
    "SN      average quality:        26.4\n",
    "SN      insert size average:    29.3\n",
    "SN      insert size standard deviation: 10.4\n",
    "SN      inward oriented pairs:  14357\n",
    "SN      outward oriented pairs: 11592\n",
    "SN      pairs with other orientation:   2819\n",
    "SN      pairs on different chromosomes: 0\n",
    "```\n",
    "\n",
    "### Try reads with BX tags and high quality\n",
    "\n",
    "NA12878_WGS read counts:\n",
    "\n",
    "```\n",
    "unmapped_or_mate_is_unmapped/NA12878_WGS.bam:  20,876,086 sequences\n",
    "samtools view -c                               21,750,049  \n",
    "```\n",
    "\n",
    "Can we merge all bins from `bxtools split`?\n",
    "\n",
    "```\n",
    "cd unmapped_or_mate_is_unmapped\n",
    "bxtools split NA12878_WGS.bam -a bx_split/NA12878_WGS -m 1000 > bx_split/NA12878_WGS_bx_counts.txt\n",
    "cd bx_split\n",
    "parallel -j30 \"bedtools bamtofastq -i {} -fq {/.}.R1.fq -fq2 {/.}.R2.fq 2>unpaired/{/.}.unpaired\" ::: *-1.bam\n",
    "cd ..\n",
    "cat bx_split/*R1.fq > NA12878_WGS.bx_only.R1.fq\n",
    "cat bx_split/*R2.fq > NA12878_WGS.bx_only.R2.fq\n",
    "```\n",
    "\n",
    "Getting total 6_401_804 out of 10_438_043 read pairs - looks quite low.\n",
    "\n",
    "Aligning to viral fasta:\n",
    "\n",
    "```\n",
    "<see commands above>\n",
    "# Real time: 36.679 sec; CPU: 361.046 sec\n",
    "\n",
    "name   mapped  coverage         divide by 809965400/1000000000 = 0.8099654 billion reads in genome\n",
    "HPV71   23552  2.93044668409    3.6179899587\n",
    "HPV47   10623  1.37496764173\n",
    "HPV82    7395  0.939644218551\n",
    "HPV42    4697  0.593280282935\n",
    "HPV51    4107  0.52599897541\n",
    "HPV105   1973  0.257336637537\n",
    "HPV20    1785  0.230114735078\n",
    "HPV102   1522  0.188412973508\n",
    "HPV21    1372  0.176372284355\n",
    "CMV     40204  0.170611849978\n",
    "```\n",
    "\n",
    "And to HPV71 only:\n",
    "0\n",
    "```\n",
    "<see commands above>\n",
    "# HPV71  30645  3.81298992161\n",
    "```\n",
    "\n",
    "Unfortunately, getting a much lower coverage from BX reads (3,8 vs 12,7).\n",
    "\n",
    "Perhaps we shouldn't be relying on bxtools stats method to get reads with BX. It's weird that summing up values from `bxtools stats` do not match the value above (`6401804 * 2 = 12_803_608` vs `19_613_313` by `cut -f2 NA12878_WGS.bam_bx_stats.txt | awk '{s+=$1} END {print s}'`)\n",
    "\n",
    "Trying to use pysam to filter reads with BX:\n",
    "\n",
    "```\n",
    "mkdir with_bx ; mkdir with_bx_lng_hqual\n",
    "./filter_bam.py NA12878_WGS.bam\n",
    "\n",
    "Total:                  21,750,049\n",
    "bx_i:                   19,613,313\n",
    "lng_i:                  20,876,724\n",
    "hqual_i:                 5,639,761\n",
    "bx_hqual_i:              5,202,715\n",
    "lng_hqual_i:             4,998,924\n",
    "bx_lng_hqual_i:          4,632,339\n",
    "paired_i:               21,750,049\n",
    "paired_hqual_i:          5,639,761\n",
    "paired_lng_hqual_i:      4,998,924\n",
    "paired_bx_hqual_i:       5,202,715\n",
    "paired_bx_lng_hqual_i:   4,632,339\n",
    "\n",
    "cd with_bx\n",
    "samtools fastq -f 1 -F 1536 NA12878_WGS.bam -1 NA12878_WGS.R1.fq -2 NA12878_WGS.R2.fq -s NA12878_WGS.single.fq\n",
    "NA12878_WGS.R1.fq      - 1,083,568 reads\n",
    "NA12878_WGS.R2.fq      - 1,083,568 reads\n",
    "NA12878_WGS.single.fq - 15,865,113 reads\n",
    "\n",
    "cd with_bx_lng_hqual\n",
    "samtools fastq -f 1 -F 1536 NA12878_WGS.bam -1 NA12878_WGS.R1.fq -2 NA12878_WGS.R2.fq -s NA12878_WGS.single.fq\n",
    "NA12878_WGS.R1.fq -         56,907 reads \n",
    "NA12878_WGS.R2.fq -         56,907 reads\n",
    "NA12878_WGS.single.fq -  4,264,669 reads  \n",
    "```\n",
    "\n",
    "Trying aligning to viral data again. With BX only:\n",
    "\n",
    "```\n",
    "HPV71   3622  0.45066567127        0.5555792877\n",
    "HPV82   1825  0.231893265565\n",
    "HPV47   1378  0.178358788506\n",
    "HPV51    859  0.110015368852\n",
    "HPV105   706  0.0920829529151\n",
    "HPV42    684  0.0863963622584\n",
    "HPV102   604  0.0747709829166\n",
    "HPV6     308  0.0385192596298\n",
    "KSHV    5184  0.0375736578507\n",
    "HPV21    281  0.0361228949736\n",
    "```\n",
    "\n",
    "With BX and long and high quality:\n",
    "\n",
    "```\n",
    "HPV71   160  0.019907925843        0.0248849073\n",
    "HPV82   155  0.0196950444727\n",
    "HPV6    125  0.0156328164082\n",
    "HPV102   90  0.0111413716266\n",
    "HPV42    58  0.00732600732601\n",
    "HPV51    50  0.00640368852459\n",
    "HPV34    26  0.00336656739609\n",
    "HPV32    24  0.00301469664615\n",
    "HPV59    20  0.00253292806484\n",
    "HPV68    18  0.00230120173869\n",
    "```\n",
    "\n",
    "Long and high quality only:\n",
    "\n",
    "```                                \n",
    "HPV71   186  0.0231429637925       0.02857156024\n",
    "HPV82   161  0.020457433291\n",
    "HPV6    146  0.0182591295648\n",
    "HPV102  110  0.0136172319881\n",
    "HPV51    64  0.00819672131148\n",
    "HPV42    64  0.00808387015284\n",
    "HPV34    30  0.00388450084164\n",
    "HPV68    26  0.00332395806699\n",
    "HPV32    26  0.00326592136666\n",
    "HPV73    20  0.0025974025974\n",
    "```\n",
    "\n",
    "Getting highly mapped reads from the HPV71 bin.\n",
    "\n",
    "With BX only:\n",
    "\n",
    "```\n",
    "HPV71  2529  0.314669652856  \n",
    "```\n",
    "\n",
    "With BX and high quality:\n",
    "\n",
    "```\n",
    "HPV71  20  0.00248849073037  \n",
    "```\n",
    "\n",
    "With BX and long and high quality:\n",
    "\n",
    "```\n",
    "HPV71  0         \n",
    "```\n",
    "\n",
    "With BX only:\n",
    "\n",
    "```\n",
    "HPV71  24  0.00298618887645  \n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Viruses in COLO829\n",
    "\n",
    "Merge fastq\n",
    "\n",
    "```\n",
    "cd /g/data3/gx8/projects/Saveliev_10X/COLO829-10x/insertions/unmapped_or_mate_is_unmapped/fastq\n",
    "gunzip -c COLO829-10x_S1_L001_R1_001.fastq.gz >> COLO829-10x_merged.fq\n",
    "gunzip -c COLO829-10x_S1_L001_R2_001.fastq.gz >> COLO829-10x_merged.fq\n",
    "  COLO829-10x_merged.fq - 121,428,506 reads\n",
    "```\n",
    "\n",
    "Searching viral sequences using 3 different approaches.\n",
    "\n",
    "- Use Mash\n",
    "```\n",
    "mash sketch /g/data3/gx8/extras/vlad/bcbio/genomes/Hsapiens/GRCh37/viral/gdc-viral.fa -i -o mash/gdc-viral\n",
    "mash screen mash/gdc-viral.msh COLO829-10x_merged.fq | sort -gr > mash/screen_gdc.tab\n",
    "head mash/screen_gdc.tab\n",
    "```\n",
    "\n",
    "Also against RefSeq:\n",
    "\n",
    "```\n",
    "mash screen refseq.genomes.k21s1000.msh ../NA12878_WGS.merged.fq | sort -gr > screen_refseq.tab\n",
    "```\n",
    "\n",
    "- Align against the viral database:\n",
    "\n",
    "```\n",
    "bwa mem -t 9 /g/data3/gx8/extras/vlad/bcbio/genomes/Hsapiens/GRCh37/viral/gdc-viral.fa COLO829.R1.fq COLO829.R2.fq | bamsort inputthreads=9 outputthreads=9 inputformat=sam index=1 indexfilename=viral_mapping/COLO829_viral.bam.bai O=viral_mapping/COLO829_viral.bam\n",
    "\n",
    "samtools idxstats viral_mapping/COLO829_viral.bam | py -x \"'\\t'.join([x.split()[0], x.split()[1], x.split()[2], str(int(x.split()[2]) / int(x.split()[1]))]) if int(x.split()[1]) > 0 else None\" | sort -nr -k4,4 | head | cols\n",
    "\n",
    "name   mapped  coverage        divide by 2033578410/1000000000 = 2.03 billion reads in genome\n",
    "HPV71  268404  33.3960432997   16.451252857\n",
    "HPV47  170086  22.0147553715\n",
    "HCV-2  178789  18.4109772423\n",
    "HCV-1  110412  11.4464026539\n",
    "HPV82   87949  11.1752223634\n",
    "HPV42   25999  3.28394593912\n",
    "HPV20   25253  3.25551115122\n",
    "HIV-1   28373  3.09040409541\n",
    "HPV73   20907  2.71519480519\n",
    "HPV21   20254  2.60367656511\n",
    "```\n",
    "\n",
    "- Filtering reads\n",
    "\n",
    "```\n",
    "./filter_bam.py COLO829-10x.bam\n",
    "\n",
    "reads_without_quality:  1,664,954\n",
    "Total:                128,214,100\n",
    "bx_i:                 112,376,814\n",
    "lng_i:                 99,917,997\n",
    "hqual_i:               83,739,552\n",
    "lng_hqual_i:           57,912,594\n",
    "bx_lng_hqual_i:        51,738,038\n",
    "paired_i:             126,549,146\n",
    "paired_lng_hqual_i:    57,912,594\n",
    "paired_bx_lng_hqual_i: 51,738,038\n",
    "\n",
    "cd with_bx_lng_hqual\n",
    "samtools fastq -f 1 -F 1536 COLO829-10x.bam -1 COLO829.R1.fq -2 COLO829.R2.fq -s COLO829.single.fq\n",
    "\n",
    "seq     len  reads  coverage         divide by 2033578410/1000000000 = 2.03 billion reads in genome\n",
    "HPV71  8037  70862  8.81697150678    4.3356929162\n",
    "HPV47  7726  30614  3.9624644059\n",
    "HCV-2  9711  36841  3.7937390588\n",
    "HPV82  7870  26405  3.35514612452\n",
    "HCV-1  9646  19666  2.03877254821\n",
    "HIV-1  9181  12356  1.34582289511\n",
    "HPV42  7917   9477  1.19704433498\n",
    "HPV51  7808   7293  0.934042008197\n",
    "HPV20  7757   6415  0.826994972283\n",
    "```\n",
    "\n",
    "- Only HPV71\n",
    "\n",
    "```\n",
    "cd /g/data3/gx8/projects/Saveliev_10X/COLO829-10x/insertions/unmapped_or_mate_is_unmapped/with_bx_lng_hqual\n",
    "samtools faidx /g/data3/gx8/extras/vlad/bcbio/genomes/Hsapiens/GRCh37/viral/gdc-viral.fa HPV71 > viral_mapping/HPV71.fa\n",
    "\n",
    "bwa index viral_mapping/HPV71.fa\n",
    "\n",
    "bwa mem -t 10 viral_mapping/HPV71.fa COLO829.R1.fq COLO829.R2.fq | bamsort inputthreads=10 outputthreads=10 inputformat=sam index=1 indexfilename=viral_mapping/COLO829_HPV71.bam.bai O=viral_mapping/COLO829_HPV71.bam\n",
    "\n",
    "samtools idxstats viral_mapping/COLO829_HPV71.bam | py -x \"'\\t'.join([x.split()[0], x.split()[2], str(int(x.split()[2]) / int(x.split()[1])), str(int(x.split()[3]) / int(x.split()[1]))]) if int(x.split()[1]) > 0 else None\" | sort -nr -k3,3 | head | cols\n",
    "\n",
    "seq    reads  coverage       divide by 2033578410/1000000000 = 2.03 billion reads in genome\n",
    "HPV71  91356  11.3669279582  5.5994718966\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Viruses in DiploidNeverResponder\n",
    "\n",
    "```\n",
    "./filter_bam.py diploid_tumor-unmapped_or_mate.unsorted.bam\n",
    "Total:              28,564,510 out of 2,230,924,710\n",
    "lng_i:              27,894,774\n",
    "hqual_i:            15,326,411\n",
    "lng_hqual_i:        14,878,067\n",
    "paired_i:           28,564,510\n",
    "paired_lng_hqual_i: 14,878,067\n",
    "\n",
    "cd lng_hqual\n",
    "samtools fastq -f 1 -F 1536 diploid_tumor-unmapped_or_mate.unsorted.bam -1 diploid_tumor.R1.fq -2 diploid_tumor.R2.fq -s diploid_tumor.single.fq\n",
    "\n",
    "#############\n",
    "### Mash\n",
    "mkdir mash\n",
    "mash sketch /g/data3/gx8/extras/vlad/bcbio/genomes/Hsapiens/GRCh37/viral/gdc-viral.fa -i -o mash/gdc-viral\n",
    "mash screen mash/gdc-viral.msh diploid_tumor.R1.fq diploid_tumor.R2.fq | sort -gr > mash/screen_gdc.tab\n",
    "head mash/screen_gdc.tab\n",
    "\n",
    "0.99726   944/1000   100  0             HPV18\n",
    "0.879223  67/1000     49  4.56495e-199  HPV45\n",
    "0.876649  63/1000     29  1.50084e-185  HPV97\n",
    "0.835578  23/1000      5  1.51569e-58   HPV59\n",
    "0.816057  14/1000     91  3.39476e-33   HPV68\n",
    "0.816057  14/1000      6  3.39476e-33   HPV85\n",
    "0.816057  14/1000   3712  3.39476e-33   HPV39\n",
    "0.816057  14/1000      2  3.39476e-33   HPV20\n",
    "0.806739  11/1000      4  3.07798e-25   HPV70\n",
    "0.806739  11/1000      4  3.07798e-25   HPV66\n",
    "\n",
    "### RefSeq\n",
    "mash screen mash/refseq.genomes.k21s1000.msh diploid_tumor.R1.fq diploid_tumor.R2.fq | sort -gr > mash/screen_refseq.tab\n",
    "(took 4:14.57 time with 100% CPU total)\n",
    "\n",
    "0.996652  932/1000  98  0             GCF_000865665.1_ViralProj15506_genomic.fna.gz   NC_001357.1 Human papillomavirus - 18, complete genome\n",
    "0.97558   595/1000   1  0             GCF_000819615.1_ViralProj14015_genomic.fna.gz   NC_001422.1 Enterobacteria phage phiX174 sensu lato, complete genome\n",
    "0.932215  229/1000   1  0             GCF_000913595.1_ViralProj222261_genomic.fna.gz  NC_022518.1 Human endogenous retrovirus K113 complete genome\n",
    "0.916976  162/1000   2  0             GCF_000287295.1_ASM28729v1_genomic.fna.gz       NC_018417.1 Candidatus Carsonella ruddii HT isolate Thao2000, complete genome\n",
    "0.887203  81/1000    2  3.28623e-247  GCF_000829155.1_ASM82915v1_genomic.fna.gz       NZ_AP013293.1 Candidatus Sulcia muelleri PSPU DNA, complete genome\n",
    "\n",
    "#############\n",
    "### BWA\n",
    "\n",
    "mkdir viral_mapping\n",
    "bwa mem -t 20 /g/data3/gx8/extras/vlad/bcbio/genomes/Hsapiens/GRCh37/viral/gdc-viral.fa diploid_tumor.R1.fq diploid_tumor.R2.fq | bamsort inputthreads=20 outputthreads=20 inputformat=sam index=1 indexfilename=viral_mapping/diploid_tumor_viral.bam.bai O=viral_mapping/diploid_tumor_viral.bam\n",
    "\n",
    "samtools idxstats viral_mapping/diploid_tumor_viral.bam | py -x \"'\\t'.join([x.split()[0], x.split()[1], x.split()[2], str(int(x.split()[2]) / int(x.split()[1]))]) if int(x.split()[1]) > 0 else None\" | sort -nr -k4,4 | head | cols\n",
    "\n",
    "name     len  mapped  coverage           divide by 1173944366/1000000000 = 1.174 billion reads in genome\n",
    "HPV18   7857  122251  15.5595010818      13.2529812606\n",
    "HPV71   8037    1714  0.213263655593\n",
    "HCV-2   9711     519  0.0534445474205\n",
    "HCV-1   9646     448  0.0464441219158\n",
    "HPV19   7685     175  0.0227716330514\n",
    "HPV82   7870     160  0.0203303684879\n",
    "HPV21   7779      46  0.00591335647255\n",
    "HPV20   7757      34  0.00438313781101\n",
    "\n",
    "#############\n",
    "### Only HPV18\n",
    "cd /g/data3/gx8/projects/Saveliev_Diploid/lng_hqual\n",
    "samtools faidx /g/data3/gx8/extras/vlad/bcbio/genomes/Hsapiens/GRCh37/viral/gdc-viral.fa HPV18 > viral_mapping/HPV18.fa\n",
    "\n",
    "bwa index viral_mapping/HPV18.fa\n",
    "\n",
    "bwa mem -t 10 viral_mapping/HPV18.fa diploid_tumor.R1.fq diploid_tumor.R2.fq | bamsort inputthreads=10 outputthreads=10 inputformat=sam index=1 indexfilename=viral_mapping/diploid_HPV18.bam.bai O=viral_mapping/diploid_HPV18.bam\n",
    "\n",
    "samtools idxstats viral_mapping/diploid_HPV18.bam | awk 'BEGIN {OFS=\"\\t\"} {print $1, $2, $3, ($2 != 0) ? $3/$2 : 0 }'\n",
    "# HPV18  122256  15.560137457\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reconstruct fastq from BAM correctly (taking trimmed barcodes into account)\n",
    "\n",
    "See discussion: \n",
    "https://github.com/samtools/hts-specs/issues/270\n",
    "\n",
    "COLO829 - HPV71\n",
    "\n",
    "```\n",
    "cd /g/data3/gx8/projects/Saveliev_10X/COLO829-10x/insertions/unmapped_or_mate_is_unmapped/with_bx_lng_hqual/viral_mapping\n",
    "```\n",
    "\n",
    "R1: 16bp (barcode) + 7bp\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filtering reads\n",
    "\n",
    "Now filtering reads even down: \n",
    "\n",
    "1. Filter low base quality and bad barcodes\n",
    "\n",
    "2. After viral alignment, filter reads with low MQ\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Assemble diploid - HPV18\n",
    "\n",
    "#### Assembly\n",
    "\n",
    "Getting HPV18 reads:\n",
    "\n",
    "```\n",
    "cd /g/data3/gx8/projects/Saveliev_Diploid/lng_hqual/viral_mapping\n",
    "\n",
    "sambamba view -F \"not unmapped or not mate_is_unmapped\" -f bam -o diploid_HPV18.one_mapped.bam diploid_HPV18_bwa.bam\n",
    "# will fail with segmentation fault, however will generate the output \n",
    "\n",
    "# !!! make sure to name-sort reads before feeding into samtools fastq - otherwise it will be ruined\n",
    "samtools sort -n diploid_HPV18.one_mapped.bam -O bam -o diploid_HPV18.one_mapped.name_sorted.bam\n",
    "\n",
    "samtools fastq diploid_HPV18.one_mapped.name_sorted.bam -1 diploid_HPV18.R1.fq -2 diploid_HPV18.R2.fq -s diploid_HPV18.single.fq\n",
    "[M::bam2fq_mainloop] discarded 0 singletons\n",
    "[M::bam2fq_mainloop] processed 119474 reads\n",
    "```\n",
    "\n",
    "Running SPAdes:\n",
    "\n",
    "```\n",
    "spades.py --only-assembler -1 diploid_HPV18.R1.fq -2 diploid_HPV18.R2.fq -s diploid_HPV18.single.fq -o spades\n",
    "\n",
    "quast.py spades/contigs.fasta -R HPV18.fa -o spades/quast --ref-bam diploid_HPV18.one_mapped.bam --no-read-stats --no-sv -1 diploid_HPV18.R1.fq -2 diploid_HPV18.R1.fq --debug\n",
    "\n",
    "# Align contigs to assembly:\n",
    "minimap2 -a HPV18.fa spades/contigs.fasta | samtools sort > spades/contigs_to_HPV18.bam && samtools index spades/contigs_to_HPV18.bam\n",
    "```\n",
    "\n",
    "----\n",
    "\n",
    "![quast](assemble/diploid/img/quast_HPV18.png)\n",
    "\n",
    "----\n",
    "\n",
    "![icarus](assemble/diploid/img/icarus_HPV18.png)\n",
    "\n",
    "----\n",
    "\n",
    "![igv](assemble/diploid/img/igv_HPV18.png)\n",
    "\n",
    "----\n",
    "\n",
    "Reads getting assembled pretty well into 2 long contigs: NODE_1 is amplified heavily up to 5kX coverage, and NODE_2 has a much smaller coverage. Contigs fully cover the viral genome, and it's clearly circular, which is visible by supplementary alignments and reverse-oriented pairs spanning the breakpoint.\n",
    "\n",
    "#### Looking for integration sites\n",
    "\n",
    "To identify if the virus is integrated into human, we need to review the viral read whose mates map to human. We would expect reads of forward orienation to pile up to the left of the breakpoint, and reads of the opposite orientation to pile up to the right of the breakpoint; the simmetrical picture would be expected in the human chromosome. In order to figure that out, we are making a GRCh37-HPV18 spike reference, and remap reads against this artificial genome:\n",
    "\n",
    "```\n",
    "mkdir GRCh37_HPV18\n",
    "cat /g/data3/gx8/local/development/bcbio/genomes/Hsapiens/GRCh37/seq/GRCh37.fa > GRCh37_HPV18/GRCh37_HPV18.fa\n",
    "cat HPV18.fa >> GRCh37_HPV18/GRCh37_HPV18.fa\n",
    "\n",
    "# Align\n",
    "bwa index GRCh37_HPV18/GRCh37_HPV18.fa\n",
    "bwa mem -t 10 GRCh37_HPV18/GRCh37_HPV18.fa diploid_HPV18.R1.fq diploid_HPV18.R2.fq | bamsort inputthreads=10 outputthreads=10 inputformat=sam index=1 indexfilename=GRCh37_HPV18/diploid_GRCh37_HPV18.bam.bai O=GRCh37_HPV18/diploid_GRCh37_HPV18.bam\n",
    "\n",
    "OR EVEN (doesn't require index):\n",
    "minimap2 -ax sr GRCh37_HPV18/GRCh37_HPV18.fa diploid_tumor.R1.fq diploid_tumor.R2.fq | bamsort inputformat=sam index=1 indexfilename=diploid_GRCh37_HPV18_minimap2.bam.bai O=diploid_GRCh37_HPV18_minimap2.bam\n",
    "```\n",
    "\n",
    "Viewing that in IGV sorted by the chromosome of mate:\n",
    "\n",
    "![igv](assemble/diploid/img/igv_spike_HPV18.png)\n",
    "\n",
    "There is no very large support for integration, however a bit of evidence from violet (chr8) and green (chr3): reads cluster in viral genome and their mates cluster in human. (Blue reads are just with too small IS and both mates map to HPV18, see https://software.broadinstitute.org/software/igv/interpreting_insert_size)\n",
    "\n",
    "Left violet reads - mapping to 8:128,316,023-128,323,877:\n",
    "\n",
    "![igv](assemble/diploid/img/igv_spike_integration_site_chr8_left.png)\n",
    "\n",
    "Right violet reads - mapping to 8:128,300,065-128,307,919:\n",
    "\n",
    "![igv](assemble/diploid/img/igv_spike_integration_site_chr8_right.png)\n",
    "\n",
    "It's possible that viruses the contig 1  the virus got inserted into chromosome 8, maybe replacing a large 15kb chunk of it.\n",
    "\n",
    "Also, green reads map to 3:186,691,636-186,699,490:\n",
    "\n",
    "![igv](assemble/diploid/img/igv_spike_chr3.png)\n",
    "\n",
    "**TODO**: exploring other viral sequences to confirm - spiking with the whole GDC database and realigning - to see if the virus integrates into 2 different chromosomes, so these reads might be coming from other viruses relative to it.\n",
    "\n",
    "It's evident that those reads are not piled up exactly in the end of contigs, but close. The explanation is that the tip of the contig is in fact a part of the genome asssembly rather than the part of the virus itself. Indeed, aligning HPV18 reference sequences back to the assembly:\n",
    "```\n",
    "minimap2 -ax asm5 spades/contigs.fasta HPV18.fa | samtools sort > spades/HPV18_to_contigs.bam && samtools index spades/HPV18_to_contigs.bam\n",
    "\n",
    "# Also map reads to the assembly:\n",
    "bwa index spades/contigs.fasta\n",
    "bwa mem -t 10 spades/contigs.fasta diploid_HPV18.R1.fq diploid_HPV18.R2.fq | bamsort inputthreads=10 outputthreads=10 inputformat=sam index=1 indexfilename=spades/reads_to_contigs.bam.bai O=spades/reads_to_contigs.bam\n",
    "```\n",
    "\n",
    "And it's evident that the assembly spans beyond HPV18 genome:\n",
    "\n",
    "![igv](assemble/diploid/img/igv_HPV18_to_contigs.png)\n",
    "\n",
    "Exploring one of the reads that map beyond HPV18 which mate maps to the assembly - `A00130:36:H32JNDSXX:2:2422:21097:19742`:\n",
    "\n",
    "![igv](assemble/diploid/img/igv_HPV18_to_contigs_mate_in_detail.png)\n",
    "\n",
    "Its mate maps normally to the assembly, but the read itself maps to GRCh37 8:128,319,919-128,320,282. Exploring the other tip of the conitg - it maps to the other chr8 location 8:128,300,065-128,307,919, supporting 2 breakpoints.\n",
    "\n",
    "Additionally blasting the contigs against human sequences, and indeed we are getting 18% identity with human chr8 alternative assembly, as well as GRCh38:\n",
    "\n",
    "![blast](assemble/diploid/img/blast_HPV18.png)\n",
    "\n",
    "#### chr8 integration site\n",
    "\n",
    "It's interesting that the _left_ HPV18 breakpoint corresponds to the _right_ chr8 breakpoint, and the _right_ HPV18 breakpoint corresponds to the left chr8 breakpoint. It suggests that it's not a simple indel, but a more complex event. Exploring the full 15kb GRCh37 region covering both breakpoints:\n",
    "```\n",
    "~/bin/sambamba slice 8:128318910-128321289 diploid_tumor-ready.bam > diploid_tumor-chr8_HPV18.bam\n",
    "```\n",
    "\n",
    "![igv](assemble/diploid/img/igv_GRCh37_full.png)\n",
    "\n",
    "We can see that the breakpoint positions are very clear, and it's also evident that:\n",
    "\n",
    "- The region is amplified heavily instead of being deleted, meaning that the virus didn't replace the region with itself,\n",
    "- The breakpoints tell us that the viruses \"left\" end got attached to the \"rightmost\" chr8 breakpoint, then looped around and attached wth it's \"right\" end to the \"leftmost\" chr8 breakpoint.\n",
    "\n",
    "It suggests that the virus created a loop that amplified the 15kb region multiple times. Indeed, things like that happen with HPV viruses, however in differrent genomic locations:\n",
    "\n",
    "![igv](assemble/diploid/img/HPV_loop_from_paper.png)\n",
    "\n",
    "The integration site overlaps long non-coding RNA genes CASC21 (Cancer Susceptibility 21, CARLo-2) and CASC8 (Cancer Susceptibility 8, CARLo-1) in their introns. Both genes are associated with cancer.\n",
    "\n",
    "8q24.21 is known as HPV integration site hotspot: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4695887/, https://www.ncbi.nlm.nih.gov/pubmed/1649348/: \"We have determined the chromosomal localization of integrated HPV type 16 (HPV-16) or HPV-18 genomes in genital cancers... In three cancers, HPV sequences were localized in chromosome band 8q24.1, in which the c-myc gene is mapped... In three of the four cases, the proto-oncogene located near integrated viral sequences was found to be structurally altered and/or overexpressed. These data indicate that HPV genomes are preferentially integrated near myc genes in invasive genital cancers and support the hypothesis that integration plays a part in tumor progression via an activation of cellular oncogenes.\", https://www.nature.com/articles/1207006: \"RSâ€“PCR of HPV18-positive tumors revealed a single large cluster at 8q24.\"\n",
    "\n",
    "\n",
    "#### chr3 integration site\n",
    "\n",
    "```\n",
    "~/bin/sambamba slice diploid_tumor-ready.bam 3:186,691,636-186,699,490 > diploid_tumor-chr3_HPV18.bam\n",
    "```\n",
    "\n",
    "![igv](assemble/diploid/img/igv_GRCh37_chr3_integration_site.png)\n",
    "\n",
    "Not clear about this one.\n",
    "\n",
    "\n",
    "\n",
    "---\n",
    "\n",
    "Aligning HPV18.fa to GRCh37.fa and GRCh38.fa:\n",
    "```\n",
    "minimap2 -ax asm5 /g/data3/gx8/extras/vlad/bcbio/genomes/Hsapiens/hg38/seq/GRCh37.fa HPV18.fa | samtools sort > HPV18_to_GRCh37.bam\n",
    "minimap2 -ax asm5 /g/data3/gx8/extras/vlad/bcbio/genomes/Hsapiens/hg38/seq/hg38.fa HPV18.fa | samtools sort > HPV18_to_hg38.bam\n",
    "```\n",
    "\n",
    "TODO: check SV in this genome location\n",
    "\n",
    "TODO: move to 10x data\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Assemble COLO829 - 10x\n",
    "\n",
    "- extract 10x reads correctly\n",
    "\n",
    "- use ema to align\n",
    "\n",
    "- Try Athena https://github.com/abishara/athena_meta\n",
    "\n",
    "- Try supernova\n",
    "\n",
    "- Visualize mapping to the virus\n",
    "\n",
    "- Visualize mapping to all viruses, with barcodes\n",
    "\n",
    "- Visualize Spades with Icarus\n",
    "\n",
    "COLO829 - HPV71\n",
    "\n",
    "/g/data3/gx8/projects/Saveliev_10X/COLO829-10x/insertions/unmapped_or_mate_is_unmapped/with_bx_lng_hqual/viral_mapping\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
